lms/djangoapps/grades/models.py

```diff
@@ -0,0 +1,269 @@
+"""
+Models used for robust grading.
+
+Robust grading allows student scores to be saved per-subsection independent
+of any changes that may occur to the course after the score is achieved.
+"""
+
+from base64 import b64encode
+from collections import namedtuple
+from hashlib import sha256
+import json
+import logging
+from operator import attrgetter
+
+from django.db import models
+from django.db.utils import IntegrityError
+from model_utils.models import TimeStampedModel
+
+from coursewarehistoryextended.fields import UnsignedBigIntAutoField
+from xmodule_django.models import CourseKeyField, UsageKeyField
+
+
+log = logging.getLogger(__name__)
+
+
+# Used to serialize information about a block at the time it was used in
+# grade calculation.
+BlockRecord = namedtuple('BlockRecord', ['locator', 'weight', 'max_score'])
+
+
+class BlockRecordList(list):
+    """
+    An ordered collection of BlockRecord objects.
+    """
+    def to_json(self):
+        """
+        Return a JSON-serialized version of the list of block records, using a
+        stable ordering.
+        """
+        sorted_blocks = sorted(self, key=attrgetter('locator'))
+        return json.dumps([block._asdict() for block in sorted_blocks], separators=(',', ':'), sort_keys=True)
+
+    def to_hash(self):
+        """
+        Return a hashed version of the list of block records.
+
+        This currently hashes using sha256, and returns a base64 encoded version
+        of the binary digest.  In the future, different algorithms could be
+        supported by adding a label indicated which algorithm was used, e.g.,
+        "sha1$witfkXg0JglCjW9RssWvTAveakI=".
+        """
+        return b64encode(sha256(self.to_json()).digest())
+
+
+class VisibleBlocksQuerySet(models.QuerySet):
+    """
+    A custom QuerySet representing VisibleBlocks.
+    """
+    def create_from_blockrecords(self, blocks):
+        """
+        Creates a new VisibleBlocks model object.
+
+        Argument 'blocks' should be an iterable collection of BlockRecords.
+        """
+        if not isinstance(blocks, BlockRecordList):
```

Comment:
It will be a BlockRecordList (now BlockRecordSet, to match our ordering-blind semantics), when we create it (as from `visible_blocks.blocks`), but it could be any iterable when created from outside this module.  I don't think it's very ergonomic to require external callers to coerce their data to this particular collection type, when all we really need is a collection of blocks.